### DISCUSSION TITLE
Build an AAB file ready for uploading to the Google Play Console

### DISCUSSION THREAD

**Discussion by Kwyrky** [MonoGame/MonoGame#8134] [Android]
Hi all,
so I received this mail from Google that I have to update or upload a new Android App or they will deactivate or delete my account or something. I uploaded some apps before in the pre AAB Format time, so when simple APKs where a thing :)
Now I know there is a tool for that now "bundletool" I think which can convert from APK to AAB or package it or whatever.
I am suddenly not able to build and deploy to Android Emulator using Visual Studio any more. I don't remember I have changed something but not sure.

![grafik](https://github.com/MonoGame/MonoGame/assets/21105488/35deec52-0086-41de-b7fc-976469a27d6d)

What do I need to change now to make it work again?

Can somebody with experience just tell me the steps necessary to make a release of that Android App and how I can get a AAB that is ready to upload in Google Play Console?

**Comment by AristurtleDev** [comment]
When .NET8 released, Microsoft deprecated the `net6.0-android` workload.  This is what these error messages are from.

Open your csproj file and change the `<TargetFramework>` from `net6.0-android` to `net8.0-android`

**Reply by Kwyrky** [reply]
I could swear I tried this some days ago and it did not work then... But now it did, thanks! :) 

I tried it this time with Visual Studio closed and used VS Code to edit the csproj file of the Android project. Not sure if this made the difference or something else went wrong last time.

So now that I can build and deploy...

I thought there where manual steps needed to get an AAB file out of it but it did automatically produce one now using Release instead of Debug.

**Reply by AristurtleDev** [reply]
Unfortunately, I don't really do any mobile development. I was just aware of how to resolve the error message you showed 😅
Hopefully someone else that does mobile will have more insight to your other questions.

**Reply by ThomasFOG** [reply]
You have to sign the aab for the store. Visual Studio (the full version, not Code) does that for you by clicking on "Archive project" (you basically just have to select your keystore file), though with VS Code, I guess you'll have to refer to google docs for signing store-ready packages: https://developer.android.com/studio/publish/app-signing

**Reply by Kwyrky** [reply]
Thanks for the help. I went through the creation of a keystore file several years ago. And I found one copy left somewhere. I tried to remember my password(s) for the keystore and the key inside of it. I will write here a bit what I tried in case anybody comes along the same issues. As a sidenote: There is Android Studio which has a GUI to create a new keystore file if you have none. But I did not want to install Android Studio necessarily since I found the old keystore file. But I could not remember the password so I continued and played the good old "guess-your-own-old-and-forgotten-password game" since I could not find it anywhere. After a while I was about to give up and create a new one but then I suddenly had an idea what I might have chosen as a password back then almost 4 years ago. And it worked! To see what is inside the keystore is btw possible WITHOUT the keystore password! Just hit "Enter" when keytool asks for the keystore password. This is what it looks like without entering any password:
`$ keytool -list -v -keystore Kwyrky.keystore`
(To make the warning about the integrity go away you can provide your password either interactively or with the storepass flag like this: `keytool -list -v -keystore Kwyrky.keystore -storepass <your_keystore_password>`)
<img width="843" alt="2024-01-27 10_51_27-Build an AAB file ready for uploading to the Google Play Console · MonoGame_Mono_EDITED" src="https://github.com/MonoGame/MonoGame/assets/21105488/c12978ac-3742-41a7-accf-dc829fdf2386">
It is helpful to know the alias for the key inside the keystore which you can find out with the last command because you need to provide the alias when configuring the signing in Visual Studio. The keystore password is needed as well for the signing process. So with this information only the password for the key which should be used in the signing process is needed. I think there is actually no direct command to verify the password for a key / alias. For security reasons I think. Interestingly enough what was possible is to run a keytool "passwd" command to change the password inside the keystore for any key inside of it! So if you know at least your keystore password you can reset your password! If you looked carefully there was also another warning at the bottom about the keystore format with a suggestion. 
![grafik](https://github.com/MonoGame/MonoGame/assets/21105488/25190978-03c1-4815-addc-3ff6554f64b6)
I found out that the passwd "trick" only works while you have the old format! 
<img width="799" alt="2024-01-27 11_36_14-Build an AAB file ready for uploading to the Google Play Console · MonoGame_Mono" src="https://github.com/MonoGame/MonoGame/assets/21105488/fef39d57-91a4-45b2-a850-8a2cedc06c72">
So I first changed the password for the keystore file with the proprietary JKS format! Only after that I migrated the keystore to the new pkcs12 industry standard format!
`$ keytool -importkeystore -srckeystore Kwyrky.keystore -destkeystore Kwyrky.keystore -deststoretype pkcs12`
<img width="963" alt="2024-01-27 11_19_32-● keytool -list -v -keystore Kwyrky keysto • Untitled-2 - Visual Studio Code" src="https://github.com/MonoGame/MonoGame/assets/21105488/11ea9f93-41cb-4d96-9938-dd4c40a1c81f">
First you are asked to enter the password for the old keystore file. If you do not provide it and just hit enter you get the warning about the integrity. After that you have to type in the password for the new keystore twice! After that it goes through the list of keys / aliases (in this case only one --> "<kwyrky>") and asks to provide the password for it so it can be imported into the new keystore.

Another info I want to leave here: 
Under the hood Visual Studio is using jarsigner which is a tool from the Java OpenJDK. If you provide a wrong password for the key or alias you will see this in the output in Visual Studio: 
> jarsigner: key associated with kwyrky not a private key
<img width="303" alt="2024-01-27 11_49_27-Firefly - Microsoft Visual Studio" src="https://github.com/MonoGame/MonoGame/assets/21105488/1e37b273-32e0-4b5e-b17b-e964a0ba0077">

An idea to try out which password works for a key if you only have the PKCS12 format keystore file is maybe to use the same jarsigner command which Visual Studio uses and there you can try to do the signing from the command line providing the key for the alias. 

So not sure if it was worth all the trouble but yeah at least now I can sign my old apps and also my new ones with the same keystore file!
